server/coordinator/procedure/operation/split/split.go

```diff
@@ -71,33 +70,65 @@ type Procedure struct {
 	state procedure.State
 }
 
-func NewProcedure(id uint64, dispatch eventdispatch.Dispatch, storage procedure.Storage, c *cluster.Cluster, schemaName string, shardID storage.ShardID, newShardID storage.ShardID, tableNames []string, targetNodeName string) *Procedure {
+func NewProcedure(id uint64, dispatch eventdispatch.Dispatch, s procedure.Storage, clusterMetadata *metadata.ClusterMetadata, snapShot metadata.Snapshot, schemaName string, shardID storage.ShardID, newShardID storage.ShardID, tableNames []string, targetNodeName string) (procedure.Procedure, error) {
+	// Validate cluster state.
+	curState := snapShot.Topology.ClusterView.State
+	if curState != storage.ClusterStateStable {
+		log.Error("cluster state must be stable", zap.Error(metadata.ErrClusterStateInvalid))
+		return nil, metadata.ErrClusterStateInvalid
+	}
+
+	found := false
+	for _, shardView := range snapShot.Topology.ShardViews {
+		if shardView.ShardID == shardID {
+			found = true
+			break
+		}
+	}
+	if !found {
+		log.Error("shard not found", zap.Uint64("shardID", uint64(shardID)), zap.Error(metadata.ErrShardNotFound))
+		return nil, metadata.ErrShardNotFound
+	}
+
+	found = false
+	for _, shardNode := range snapShot.Topology.ClusterView.ShardNodes {
+		if shardNode.ShardRole == storage.ShardRoleLeader {
+			found = true
+		}
+	}
+	if !found {
+		log.Error("shard leader not found", zap.Error(procedure.ErrShardLeaderNotFound))
+		return nil, procedure.ErrShardLeaderNotFound
+	}
```

Comment:
I guess these checks should be defined as a free function which receives only `Topology` as params, and the unit test can be easily built if so.